R/kiss-report.R

```diff
@@ -4,47 +4,57 @@
 #' @return If url is a valid string or url (from \code{httr::url}) and the start
 #'   and end dates are specified this returns a new KissReport object that can
 #'   be passed to the read S3 generic
 #' @examples
-#'    reportUrl <- "https://query.kissmetrics.com/v2/products/6581c29e-ab13-1030-97f2-22000a91b1a1/reports/1c564450-3586-0133-85e2-22000a9a8afc/run"
-#'    report <- KissReport(reportUrl,
-#'                        interval = lubridate::interval(as.Date("2015-06-01"), as.Date("2015-06-02")),
-#'                        columnNames = c("KM_Email", "KM_FirstUserId", "FirstVisitDate",
-#'                                        "FirstSource", "FirstMedium",
-#'                                        "FirstCampaignName", "FirstCampaignContent", "FirstCampaignTerms",
-#'                                        "FirstReferrer"))
+#'    reportDates <- lubridate::interval(as.Date("2015-06-01"), as.Date("2015-06-02"))
+#'    rules <- list(KissRule.Event(FALSE, 72, 1, "at_least", "any_value"))
+#'    segment <- KissSegment(type = "and",
+#'                 rules = rules,
+#'                 defaultInterval = reportDates)
+#'    reportV3 <- KissReport(productId = "6581c29e-ab13-1030-97f2-22000a91b1a1",
+#'                 segment = segment,
+#'                 calculations = list(
+#'                   KissCalculation.Event(label = "First time of visited site",
+#'                      eventId = 6,
+#'                      type = "first_date_in_range",
+#'                      negate = FALSE,
+#'                      frequencyValue = 1,
+#'                      frequencyOccurance = "at_least")),
+#'                 interval = reportDates,
+#'                 )
 #'    reportResults <- read(report)
 #' @export
-KissReport <- function(url, start, end, interval, columnNames) {
-  if (!is.character(url) && !httr::is.url(url)) stop("url must be a string or url class")
-  if (is.null(interval)) {
-    if (!lubridate::is.Date(start)) stop("If not creating with an interval, start must be a date")
-    if (!lubridate::is.Date(end)) stop("If not creating with an interval, end must be a date")
-    interval <- interval(start, end)
-  } else {
-    if (!lubridate::is.interval(interval)) stop("interval must be a valid interval")
+KissReport <- function(productId, segment, calculations, interval) {
+  if (!is.character(productId)) stop("productId must be a string")
+  if (!lubridate::is.interval(interval)) stop("interval must be a valid interval")
+  if (!is(segment, "KissSegment"))  stop("segment must be a KissSegment object")
+  if (!is.list(calculations))  stop("calculations must be a list")
+  if ( length(calculations) > 0 & !sapply(calculations, is, "KissCalculation")) {
+    stop("calculations must contain only KissCalculations")
   }
 
   cache <- new.env(parent = emptyenv())
 
-  url <- httr::parse_url(url)
+  urlTemplate <- "https://query.kissmetrics.com/v2/products/{{product_id}}/reports/people_search"
```

Comment:
Phase 2 - for loading from external file. 

R doesn't seem to have a concept of const